### DISCUSSION TITLE
Doubt regarding the GATConv

### DISCUSSION THREAD

**Discussion by arake001** [pyg-team/pytorch_geometric#2941] [Q&A]
In the below code snippet. Can you explain the scenario where x_l and x_r are different?  I think x will always be a tensor. Therefore, x_l and x_r will always be the same.

```python
if isinstance(x, Tensor):
    assert x.dim() == 2, 'Static graphs not supported in GATConv.'
    x_l = x_r = self.lin_l(x).view(-1, H, C)
    alpha_l = (x_l * self.att_l).sum(dim=-1)
    alpha_r = (x_r * self.att_r).sum(dim=-1)
else:
            x_l, x_r = x[0], x[1]
            print ('not same')
            pdb.set_trace()
            assert x[0].dim() == 2, 'Static graphs not supported in `GATConv`.'
            x_l = self.lin_l(x_l).view(-1, H, C)
            alpha_l = (x_l * self.att_l).sum(dim=-1)
            if x_r is not None:
                x_r = self.lin_r(x_r).view(-1, H, C)
                alpha_r = (x_r * self.att_r).sum(dim=-1)
```

The if statement will always be true in my opinion, as x is always a tensor. I tried with the TUDataset example. The else statement is not executed even once. Can you check or give an example where the other use case is valid? Any help would be appreciated. @rusty1s

Thank you!

**Comment by rusty1s** [comment]
The `GATConv` operator supports both message passing in homogeneous graphs (where `x` is just a regular `Tensor`), as well as message passing in bipartite graphs (where `x` is a tuple of two tensors: the node features of source and target nodes):
```python
conv = GATConv(in_channels=(64, 128), out_channels=256)

x_src = torch.randn(num_src_nodes, 64)
x_dst = torch.randn(num_dst_nodes, 128)

out = conv((x_src, x_dst), edge_index)
```
This will compute new embeddings for target nodes based on incoming features (`x_src`) from source nodes. Hope this clarifies your issues :)

**Reply by liujxing** [reply]
I also have a question about this: is there a specific order on the index of the nodes? For example if I have 

```
num_src_nodes = 2
num_dst_nodes = 3
edge_index = [[0, 0, 0, 1, 1, 1], [2, 3, 4, 2, 3, 4]]
x_src = torch.randn(num_src_nodes, 64)
x_dst = torch.randn(num_dst_nodes, 99)
out = conv((x_src, x_dst), edge_index)
```
How does the library know how to map the nodes in edge_index to nodes in features?

**Reply by rusty1s** [reply]
This is done via the indices in `edge_index`, *e.g.*, a value of `0` in `edge_index` maps to the feature value in `x[0]`.

**Reply by liujxing** [reply]
Yes that I understand, my question is more like "how does it know whether `x[0]` should be matched to `x_src[0]` or `x_dst[0]` "?

**Reply by rusty1s** [reply]
`edge_index[0]` (the source nodes) will be used to index `x_src`, while `edge_index[1]` (the destination nodes) will be used to indexed `x_dst`.

**Reply by liujxing** [reply]
now I understand, thanks a lot!
